src/collectors/proc.plugin/sys_class_power_supply.c

```diff
@@ -128,85 +111,18 @@ void power_supply_free(struct power_supply *ps) {
-        if(unlikely(prop->fd == -1)) {
-            collector_error("Cannot open file '%s'", prop->filename);
-            return;
-        }
-    }
-    ssize_t r = read(prop->fd, buffer, PROP_VALUE_LENGTH_MAX);
-    if(unlikely(r < 1)) {
-        collector_error("Cannot read file '%s'", prop->filename);
-    }
-    else {
-        buffer[r] = '\0';
-        prop->value = str2ull(buffer, NULL);
-        prop->ok = true;
-    }
-
-    if(unlikely(!keep_fds_open)) {
-        close(prop->fd);
-        prop->fd = -1;
-    }
-    else if(unlikely(prop->ok && lseek(prop->fd, 0, SEEK_SET) == -1)) {
-        collector_error("Cannot seek in file '%s'", prop->filename);
-        close(prop->fd);
-        prop->fd = -1;
-    }
-    return;
-}
-
-static void rrdset_create_simple_prop(struct power_supply *ps, struct simple_property *prop, char *title, char *dim, collected_number divisor, char *units, long priority, int update_every) {
-    if(unlikely(!prop->st)) {
-        char id[RRD_ID_LENGTH_MAX + 1], context[RRD_ID_LENGTH_MAX + 1];
-        snprintfz(id, RRD_ID_LENGTH_MAX, "powersupply_%s", dim);
-        snprintfz(context, RRD_ID_LENGTH_MAX, "powersupply.%s", dim);
-
-        prop->st = rrdset_create_localhost(
-                id
-                , ps->name
-                , NULL
-                , dim
-                , context
-                , title
-                , units
-                , PLUGIN_PROC_NAME
-                , PLUGIN_PROC_MODULE_POWER_SUPPLY_NAME
-                , priority
-                , update_every
-                , RRDSET_TYPE_LINE
-        );
-
-        add_labels_to_power_supply(ps, prop->st);
-    }
-
-    if(unlikely(!prop->rd)) prop->rd = rrddim_add(prop->st, dim, NULL, 1, divisor, RRD_ALGORITHM_ABSOLUTE);
-    rrddim_set_by_pointer(prop->st, prop->rd, prop->value);
-
-    rrdset_done(prop->st);
-}
-
 int do_sys_class_power_supply(int update_every, usec_t dt) {
     (void)dt;
-    static int do_capacity = -1, do_power = -1, do_property[3] = {-1};
+    static int do_property[5] = {-1};
     static int keep_fds_open = CONFIG_BOOLEAN_NO, keep_fds_open_config = -1;
     static char *dirname = NULL;
 
-    if(unlikely(do_capacity == -1)) {
-        do_capacity    = config_get_boolean("plugin:proc:/sys/class/power_supply", "battery capacity", CONFIG_BOOLEAN_YES);
-        do_power       = config_get_boolean("plugin:proc:/sys/class/power_supply", "battery power", CONFIG_BOOLEAN_YES);
-        do_property[0] = config_get_boolean("plugin:proc:/sys/class/power_supply", "battery charge", CONFIG_BOOLEAN_NO);
-        do_property[1] = config_get_boolean("plugin:proc:/sys/class/power_supply", "battery energy", CONFIG_BOOLEAN_NO);
-        do_property[2] = config_get_boolean("plugin:proc:/sys/class/power_supply", "power supply voltage", CONFIG_BOOLEAN_NO);
+    if(unlikely(do_property[0] == -1)) {
+        do_property[0] = config_get_boolean("plugin:proc:/sys/class/power_supply", "battery capacity", CONFIG_BOOLEAN_YES);
```

Comment:
Considering you have a specific value for each index, I suggest creating an `enum` and use them. At the end of `enum`, please, add another value that symbolizes the end to use in loops.